src/Core/ScreenshotCache.vala

```diff
@@ -0,0 +1,201 @@
+public class AppCenterCore.ScreenshotCache {
+    private const int MAX_CACHE_SIZE = 100000000;
+
+    public string screenshot_path;
+
+    // Atomic integer for keeping track of the current screenshot usage.
+    private int screenshot_usage;
+
+    public ScreenshotCache () {
+        screenshot_path = GLib.Environment.get_user_cache_dir ()
+            + "/"
+            + Build.PROJECT_NAME
+            + "/screenshots";
+
+        if (GLib.DirUtils.create_with_parents (screenshot_path, 0755) == -1) {
+            critical (
+                "Error creating the temporary folder: GFileError #%d",
+                GLib.FileUtils.error_from_errno (GLib.errno)
+            );
+        }
+
+        new Thread<bool> ("screenshot_cache", () => {
+            cache_removal_event_loop ();
+            return true;
+        });
+    }
+
+    // When `screenshot_usage` exceeds `MAX_CACHE_SIZE`, the oldest files (based on ctime) will be deleted.
+    private void cache_removal_event_loop () {
+        screenshot_usage = summarize_screenshot_usage ();
+        while (true) {
```

Comment:
Do we really have to run always here? Can we do cleanups only when actually necessary, when downloading new screenshots?

Reply:
It does only perform a cleanup when necessary, communicated via an atomic integer that's incremented from the threads that fetch new screenshots. You could combine them into an extra step that re-calculates the cache size after fetching screenshots for an app. Not sure how much longer that would make the fetching process take.